### PR TITLE
Fix sliceviewer colorbar autoscale

### LINKED ISSUE #32359: SliceViewer z-axis autoscale lag
**Describe the bug**
In the sliceviewer, autoscaling of the z(colour)-axis lags behind zoom operations by one. However, it seems to work fine for event workspaces.

**To Reproduce**
1. Load workspace and open the slice viewer.
2. Make sure "Autoscaling" is ticked (bottom right of the SliceViewer window).
3. Zoom into a region of low occupancy using the "zoom to rectangle" tool.
4. The limits of the z-axis will remain the same until you perform another zoom operation, e.g. scroll once into the plot.
5. Click the home button to reset the x- and y-axis limts.
6. The limits of the z-axis will not update to account for the range of the data until you click the home button **again**.

**Expected behavior**
z-axis should recale correctly after each zoom operation.

**Platform/Version (please complete the following information):**
 - OS: All
 - Mantid Version 6.1

**Additional context**

### LINKED ISSUE #33214: Sliceviewer: Colorbar autoscaling not behaving as expected
**Describe the bug**

During manual testing for 6.3 (#33201) it was discovered that colorbar autoscaling does not work as expected and provides an inconsistent experience for the user.

**This is a regression and must be fixed for 6.3**

**To Reproduce**

Use the data and step 1 from the [manual testing instructions](https://developer.mantidproject.org/Testing/SliceViewer/SliceViewer.html#viewing-data). From manual testing it was observed:
 - MatrixWorkspace: Zoom in with autoscale checked, hit Home and the autoscaled value does not match the value you get if you then uncheck/recheck the box.
 - MDWorkspace (2D/3D/Histo): Scale not updating as you zoom

**Expected behavior**
* While enabled the colobar scale should update to the min/max of the viewed slice.

**Platform/Version (please complete the following information):**
 - OS: Tested on Windows
 - OS Version: 10
 - Mantid Version mantidnightly-6.2.20220104.1619

### PR DESCRIPTION
**Pull request by gemmaguest**
This PR fixes two issues that are causing autoscale to fail for the sliceviewer colorbar:

- Move the `canvas.draw` to before calculating the colorbar limits, because the limits are based on the image data. For Workspace2D this is not recalculated before we redraw.
- For MD workspaces, add a new function to get the image data clipped to the axes extents (because the default `get_array` function always returns the full extent). Note that this works regardless of where the draw is done with the previous fix.

Fixes #33214, fixes #32359

**To test:**

See the instructions in the comment by Richard in the linked issue. Check that autoscaling now works as expected when you zoom in and out. 

Release notes to follow

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by gemmaguest** [comment]
@RichardWaiteSTFC Sorry for the force push. I've just:
- added release notes 
- simplified the fudge factor for finding pixel edges
- removed a bit of refactoring I did to `extract_matched_slices` which is no longer necessary

**Comment by RichardWaiteSTFC** [comment]
> @RichardWaiteSTFC Sorry for the force push. I've just:
> 
> * added release notes
> * simplified the fudge factor for finding pixel edges
> * removed a bit of refactoring I did to `extract_matched_slices` which is no longer necessary

No worries, is it ready for review now? No rush!

**Review by RichardWaiteSTFC — COMMENTED** [review]
Just done some testing on Worksapce2D functionality and found a couple of things:
1. It looks like the original colorbar limit is set from the data (all spectra and bins), but if you toggle the autoscale then it sets it using the sampled/interpolated image - it would be good if it was always set to the latter.
image
2. This is minor but resizing the sliceviewer window and toggling autoscale sometimes gives a different color limit (can be hard to reproduce) - this makes sense because the sampled image is different, but maybe we should update the colorbar on resizing if autoscale is on? Don't feel strongly about this though.

**Comment by RichardWaiteSTFC** [comment]
Functional testing on MDEvent and MDHisto workspaces 2D-4D looks good!

File: Framework/PythonInterface/mantid/plots/modest_image/modest_image.py [hunk 1]
```diff
@@ -87,6 +87,25 @@ def get_array(self):
...
+
+        # Add 0.5 to get the edge of the pixel. Also add 1 to the max values which need to be one past the end
+        # for when we slice.
+        y0 = max(int(np.floor(ind0[1] + 0.5)), 0)
+        y1 = max(int(np.floor(ind1[1] + 0.5)) + 1, 0)
+        x0 = max(int(np.floor(ind0[0] + 0.5)), 0)
+        x1 = max(int(np.floor(ind1[0] + 0.5)) + 1, 0)
+
+        # Clip the data to the extents
+        data = self._full_res[y0:y1:1, x0:x1:1]
```
Comment (RichardWaiteSTFC) [inline Framework/PythonInterface/mantid/plots/modest_image/modest_image.py:105]: Really nit-picking but we don't need to specify a step of 1 when slicing - happy to leave as is though.

File: qt/python/mantidqt/mantidqt/widgets/colorbar/colorbar.py [hunk 2]
```diff
@@ -295,7 +295,15 @@ def update_clim_validator(self):
     def _autoscale_clim(self):
         """Update stored colorbar limits
         The new limits are found from the colobar data """
-        data = self.colorbar.mappable.get_array()
+        data = self.colorbar.mappable.get_array_clipped_to_bounds()
+
+        # If any dimension is zero then we have no data in the display area
+        if any(map(lambda dim: dim == 0, data.shape)):
```
Comment (RichardWaiteSTFC) [inline qt/python/mantidqt/mantidqt/widgets/colorbar/colorbar.py:301]: This is a really good addition, I believe it works for MD Histo workspaces (I think it would work for MDEvent anyway becasue BinMD get recalled even for limits outside the extent of the data but it's hard to confirm for MDHisto workspaces as autoscale previously didn't work at all!).
However, I've found that in a Workspace2D (e.g. `MAR11060.raw` which has a spike in the last bin of every spectrum which makes this bug easier to reproduce) if you zoom outside the extent of the data the color limits don't seem to update to (cmax, cmin) = (0,0) it gets frozen at whatever the limits were when there was still data on the image. Or maybe there is a bug elsewhere?

**Review by RichardWaiteSTFC — CHANGES_REQUESTED** [review]
This is a big improvement and really good tests have been added for the modest image and sliceviewer view. It works very well for MDHisto and MDEvent objects but I have found a couple of issues with Workspace2D - which maybe indicates some edge cases for which there are no tests? I think they are quite minor so if you want to address these in a separate PR then that's fine be me and I will approve.

File: qt/python/mantidqt/mantidqt/widgets/colorbar/colorbar.py (hunk 2 above)
Comment (gemmaguest) [reply qt/python/mantidqt/mantidqt/widgets/colorbar/colorbar.py]: I've moved the check to the masked data rather than the original data which seems to fix this.

**Comment by gemmaguest** [comment]
> Just done some testing on Worksapce2D functionality and found a couple of things:
> 
>     1. It looks like the original colorbar limit is set from the data (all spectra and bins), but if you toggle the autoscale then it sets it using the sampled/interpolated image - it would be good if it was always set to the latter.
>        image
> 
>     2. This is minor but resizing the sliceviewer window and toggling autoscale sometimes gives a different color limit (can be hard to reproduce) - this makes sense because the sampled image is different, but maybe we should update the colorbar on resizing if autoscale is on? Don't feel strongly about this though.

I think both of these would be fixed by updating the colorbar on resize. I've had a look at implementing that but I think it means adding a new signal and I think it's probably more work/risk than it's worth for this release.

**Comment by RichardWaiteSTFC** [comment]
> > Just done some testing on Worksapce2D functionality and found a couple of things:
> > ```
> > 1. It looks like the original colorbar limit is set from the data (all spectra and bins), but if you toggle the autoscale then it sets it using the sampled/interpolated image - it would be good if it was always set to the latter.
> >    image
> > 
> > 2. This is minor but resizing the sliceviewer window and toggling autoscale sometimes gives a different color limit (can be hard to reproduce) - this makes sense because the sampled image is different, but maybe we should update the colorbar on resizing if autoscale is on? Don't feel strongly about this though.
> > ```
> 
> I think both of these would be fixed by updating the colorbar on resize. I've had a look at implementing that but I think it means adding a new signal and I think it's probably more work/risk than it's worth for this release.

Ah yes that would make sense! Agreed adding a new signal should wait till the next release - I will make an issue for this.

**Comment by gemmaguest** [comment]
> Ah yes that would make sense! Agreed adding a new signal should wait till the next release - I will make an issue for this.

Great, thanks.

This should be ready for re-testing now

File: qt/python/mantidqt/mantidqt/widgets/colorbar/colorbar.py [hunk 3]
```diff
@@ -296,23 +296,19 @@ def _autoscale_clim(self):
...
             try:
                 masked_data = data[~data.mask]
                 # use the smallest positive value as vmin when using log scale,
                 # matplotlib will take care of the data skipping part.
                 masked_data = masked_data[masked_data > 0] if norm == "Log" else masked_data
-                self.cmin_value = masked_data.min()
-                self.cmax_value = masked_data.max()
+
+                # If any dimension is zero then we have no data in the display area
+                data_is_empty = any(map(lambda dim: dim == 0, masked_data.shape))
```
Comment (RichardWaiteSTFC) [inline qt/python/mantidqt/mantidqt/widgets/colorbar/colorbar.py:308]: This works nicely now - I note that this doesn't cause an issue for log colorscale as when you get limits set to (0,0) then it switches the scale to linear.

**Review by RichardWaiteSTFC — APPROVED** [review]
Looks good, thanks for that last fix and tidy up!
I found one other issue with Workspace2D - which is that we should update the color axis limits when transposing data (for the same reason as we want to update them when resizing the window) - I will add this to issue #33355

**Review by martyngigg — APPROVED** [review]
Nice set of fixes and great to see big increase in test coverage
